### PR TITLE
tdnf-python: fix regression in tdnf.repolist()

### PR DESCRIPTION
**Pull request by sshedi**
repo->pszMetaLink can be null and PyBytes_FromString doesn't validate
null pointers which was resulting in a seg fault.

Other changes:

1. Added dist flag support to spec (which helps in reinstalling tdnf if
   needed)
2. Fixed a typo in python/README.md
3. Improvements to pytests/conftest.py
4. Added config file parameter support to tdnf-python

Signed-off-by: Shreenidhi Shedi <sshedi@vmware.com>

### CODE REVIEW

**Comment by sshedi** [comment]
@tapakund @oliverkurth I will add tests later. I'm still having a hard time to add tests for this.
For the time being, let's get these changes merged first.

File: pytests/conftest.py [hunk 1]
```diff
@@ -53,39 +57,31 @@ def make_cert(self):
...
-                                                    server_side = True)
-            self.httpd.serve_forever()
-        finally:
-            self.httpd.server_close()
-
-    def join(self):
-        self.httpd.shutdown()
-        super().join()
+        keyfile = os.path.join(os.getcwd(), 'key.pem')
+        certfile = os.path.join(os.getcwd(), 'cert.pem')
```
Comment (oliverkurth) [inline pytests/conftest.py]: This is odd - we set `keyfile` and `certfile` in line 45/46, then open the files *without* using the variables, then assign the same values again. I know this isn't new, but should be rewritten. Also, assigning those values to local variables in the if clause (instead of attributes to `self`) in lines 45/46 doesn't seem to make sense if we are returning.

File: pytests/conftest.py [hunk 2]
```diff
@@ -1,45 +1,49 @@
...
+def StopTestRepoServer(server):
+    server.terminate()
+    server.join()
 
-    def make_cert(self):
+def TestRepoServer(root, port=8080, interface="", enable_https=False):
+    port = port
+    root = root
+    addr = (interface, port)
+    enable_https = enable_https
```
Comment (oliverkurth) [inline pytests/conftest.py:37]: Remove the lines that are not doing anything (`foo=foo`).

File: pytests/conftest.py [hunk 3]
```diff
@@ -280,4 +274,27 @@ def _run(self, cmd, retvalonly=False, cwd=None):
...
+    server = Process(target=TestRepoServer,
+                     args=(test_utils.config['repo_path'], ))
+    server.start()
+
+    atexit.register(StopTestRepoServer, server)
+
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    retry = 0
+    while retry < 10:
+        retry += 1
```
Comment (oliverkurth) [inline pytests/conftest.py]: Nit: Consider using a `for` loop: `for retry in range(1, 10):`. Just be careful testing the value in line 296.

File: pytests/tests/test_tdnf_python.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    global sglpkgname
+    sglpkgname = utils.config['sglversion_pkgname']
+
+    global config
+    config = utils.config['repo_path'] + '/tdnf.conf'
+
+    path = utils.config['build_dir']
+    path += '/python/build/lib.linux-' + os.uname().machine + '-'
+    path += str(sys.version_info.major)+'.'+str(sys.version_info.minor)
```
Comment (oliverkurth) [inline pytests/tests/test_tdnf_python.py]: Nit: you can use `format()` here.

File: pytests/tests/test_tdnf_python.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+def teardown_test(utils):
+    utils.run(['tdnf', 'erase', '-y', sglpkgname])
+    utils.run(['tdnf', 'erase', '-y', mulpkgname])
+    assert (utils.check_package(sglpkgname) == False)
+    assert (utils.check_package(mulpkgname) == False)
+
+
+def tdnf_py_erase(pkgs, quiet=False, refresh=False):
+    pkgs = [x for x in pkgs if x]
```
Comment (oliverkurth) [inline pytests/tests/test_tdnf_python.py:51]: Why check for unset entries? Can this happen?

Comment (sshedi) [reply pytests/tests/test_tdnf_python.py:51]: Yes, for some weird reason the pkg list has entries like [b' ', 'tdnf-test-one']

File: pytests/conftest.py [hunk 6]
```diff
@@ -280,4 +265,25 @@ def _run(self, cmd, retvalonly=False, cwd=None):
...
+        result = sock.connect_ex(('localhost', 8080))
+        if not result:
+            print("Server is up and running")
+            sock.close()
+            break
+        else:
+            print("Server is not running, retrying ...")
+            time.sleep(1)
+
+    if (retry == 10):
```
Comment (oliverkurth) [inline pytests/conftest.py]: This won't tell you if it succeeded or not. It may have succeeded with the last attempt, or not. Better test `result`.

**Review by oliverkurth — COMMENTED** [review]
I am not familiar with the tdnfmodule stuff yet - I need to catch up on that.

File: pytests/conftest.py (hunk 6 above)
Comment (sshedi) [reply pytests/conftest.py]: Nice, ty. Done.

File: ci/docker-entrypoint.sh [hunk 7]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
 
 rm -rf build
-mkdir build && cd build
-cmake .. && make && make python
-make check
+mkdir -p build
+cd build || exit 1
+cmake .. && make -j32 && make python -j32
+make check -j32
```
Comment (oliverkurth) [inline ci/docker-entrypoint.sh:7]: I like that the build is much faster with `-j32`, but I don't think it has an effect for `make check`.

**Review by oliverkurth — APPROVED** [review]
Looks good to me.

I would reword the commit message because it does more that what it says, or split it into multiple commits per issue addressed - this would make it easier to revert in case something breaks.

I am not happy with the python API - we shouldn't need to give options to every method call, instead this should be set once in a context class. But that's another topic.

File: ci/docker-entrypoint.sh (hunk 7 above)
Comment (sshedi) [reply ci/docker-entrypoint.sh:7]: I know. I just kept it anyway, assuming it might do something under the hood :)
It's not causing any issues, I think it's okay to keep it.

**Comment by sshedi** [comment]
> Looks good to me.
> 
> I would reword the commit message because it does more that what it says, or split it into multiple commits per issue addressed - this would make it easier to revert in case something breaks.
> 
> I am not happy with the python API - we shouldn't need to give options to every method call, instead this should be set once in a context class. But that's another topic.

I changed my commit message a bit. Thanks.
And yes, python api framework is not looking good, I just made it work with existing code.
If any real requirement comes up we can refactor it.
